File: Difficulty: Medium/Kth Ancestor in a Tree/kth-ancestor-in-a-tree.cpp
/*
Structure of the node of the binary tree is as
struct Node
{
    int data;
    struct Node *left, *right;
};
*/
// your task is to complete this function
class Solution {
  public:
    int dfs(Node* root,int& k,int node,int level,int& backLevel,int& ans) {
        if(root==nullptr) return -1;
        if(root->data == node ) {
            //k=k-1;
            //backLevel++;
            return level;
        }
        int left = dfs(root->left,k,node,level+1, backLevel,ans);
        if(left!=-1) {
            backLevel++;
            if(backLevel==k) {
                ans = root->data;
            }
            return left;
        }
        int right = dfs(root->right,k,node,level+1,backLevel,ans);
          if(right!=-1) {
            backLevel++;
            if(backLevel==k) {
                ans = root->data;
            }
            return right;
        }
        return right;
        
    }
    int kthAncestor(Node *root, int k, int node) {
        // Code here
        int level = 0;
        int ans = -1;
        int backLevel = 0;
        dfs(root,k,node,level,backLevel ,ans);
        return ans;
    }
};

File: Difficulty: Medium/Kth Ancestor in a Tree/kth-ancestor-in-a-tree.java
/* Definition for Node
class Node {
    public int data;
    public Node left;
    public Node right;

    public Node(int val) {
        data = val;
        left = right = null;
    }
};
*/

class Solution {
    
    public int solve(Node root,int k, int node, int[] ans) {
        
        if(root==null) return Integer.MIN_VALUE;
        
        if(root.data==node) return 0;
        
        int left = solve(root.left,k,node,ans);
    
        int right = Integer.MIN_VALUE;
        
        if(left==Integer.MIN_VALUE)
            right = solve(root.right,k,node,ans);
        
        int cur = Math.max(left,right);
        
        if(cur==Integer.MIN_VALUE) return cur;
        
        cur+=1;
        
        //System.out.println(cur);
        
        if(cur==k) {
            ans[0] = root.data;
        }
        
        return cur;
        
    }
    
    public int kthAncestor(Node root, int k, int node) {
        // code here
        
        int[] ans = new int[1];
        ans[0] = -1;
        
        solve(root,k,node,ans);
        
        return ans[0];
        
    }
}